rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -1966,16 +1968,26 @@ impl SystemState {
                 return;
             }
 
-            // Protect against enqueuing a second response for the currently executing
-            // (aborted or paused) callback.
+            // Protect against enqueuing duplicate responses.
             if let RequestOrResponse::Response(response) = &msg {
-                if let Some(aborted_or_paused_response) = self.aborted_or_paused_response() {
-                    if response.originator_reply_callback
-                        == aborted_or_paused_response.originator_reply_callback
-                    {
-                        // The callback is already executing, don't enqueue a second response for it.
-                        return;
+                match should_enqueue_input(
+                    response,
+                    call_context_manager,
+                    self.aborted_or_paused_response(),
+                ) {
+                    // Safe to induct.
+                    Ok(true) => {}
+
+                    // Best effort response whose callback is gone. Silently drop it.
+                    Ok(false) => {
+                        self.queues
+                            .pop_canister_output(&self.canister_id)
+                            .expect("Message peeked above so pop should not fail.");
```

Comment:
Instead of this here, did you consider just bailing out here as for the case below? It is not a clear cut as it would be less efficient to hand things over to MR but also come with less cognitive overhead.

Reply:
The case below is something like the critical errors we are seeing now. It's likely dangerous to do anything about it midway through an execution round. Or at least cleaner to leave it to Message Routing to deal with uniformly.

This OTOH is not an error case (hence it being wrapped in an `Ok`). It's a best-effort response for a callback that's clearly gotten a response already, so it's part of the expected behavior of the protocol.

`should_enqueue_input()` returning `Ok(true)` means "safe to induct"; it returning `Ok(false)` simply means "late best-effort response, safe to drop". Anything else would be an `Err`.

Reply:
Agreed regarding the error case vs. the non error case. 

My thinking comes from the observation that we also returned in the case where a callback was aborted or paused before this change -- so in some sense also leaving the work to MR in the non-error case. So I was wondering whether we could get away with something similar now, i.e., where we don't pop messages from output queues and buy the related cognitive overhead of having to think what this means for all the invariants etc. 

I'm happy if we go ahead with this if it is your preference. It just means that I will have to look into it deeper now 😄 

Meanwhile, if we want to stick with this, are there any `debug_asserts` regarding queue invariants we  could add here?

Reply:
[As mentioned offline, I never saw this comment, whether because of a GitHub related browser extension or because of GitHub in general.]

The reason why were bailing out earlier was that the logic to bail out for duplicates of paused/aborted callbacks had been added for guaranteed responses. For which, outside of bugs, there would have been no source of responses to non-existent callbacks, so I only checked explicitly against duplicates for paused/aborted callbacks. Which would have been a bug.

For best-effort responses, responses for either no (longer) existent or paused/aborted callbacks are an expected part of the protocol, so there is no need to defer to Message Routing to deal with them.

The main change is that before the choices were to either induct the message or bail out early; now we can induct the message (if it's a request or it's a response that matches a live callback), silently drop it (if it's a best effort response not matching any live callback) or bail out (in case of an error that is outside the expected behavior of the protocol). And we use the same filter as for `push_input()` (namely `should_enqueue_input()`). I don't think that apart from the slightly increased complexity (we're now dealing with two classes of message) there's any reason for concern.

Regardless, I've added the `debug_asserts`, as they are there on all other code paths that mutate `CanisterQueues`. Thanks.